feat(category_theory): basic definitions

kim-em opened the pull request:
This is just the start of a PR of my category theory library.

I'm very happy to add components to this PR if it is easier to see more things in context.

I am not attached at all to any of the notational choices, and am quite happy to change these. (I do really want to keep a notation like the current `f ≫ g` for "do f, then g", but don't mind the symbol.)

Commit:
feat(category_theory): basic definitions

Commit:
fixing formatting in documentation

jcommelin commented on category_theory/functor/default.lean:
```diff
@@ -0,0 +1,89 @@
+variable {C : Type u₁}
+variable [category.{u₁ v₁} C]
+
+@[simp] lemma IdentityFunctor.onObjects   {C : Type u₁} [category.{u₁ v₁} C] (X : C)                 : (IdentityFunctor C) +> X = X := by refl
```
The colon doesn't algin with the line below, but it looks like you want it to.

spl replied:
It seems there are duplicate variables `{C : Type u₁} [category.{u₁ v₁} C]` here?

spl replied:
... and below here, as well.

kim-em replied:
I couldn't get it to align exactly (slight variations in character widths), so I've removed all the spurious spaces.

@spl, I don't think there was any actual duplication, because of sectioning commands, but in any case I've tidied the variables up.

jcommelin commented on category_theory/natural_transformation.lean:
```diff
@@ -0,0 +1,111 @@
+
+@[simp,ematch] lemma vertical_composition_of_NaturalTransformations.components (α : F ⟹ G) (β : G ⟹ H) (X : C) : (α ⊟ β).components X = (α.components X) ≫ (β.components X) := by refl
+
+-- We'll want to be able to prove that two natural transformations are equal if they are componentwise equal.
```
It seems like this is an intermezzo between vertical and horizontal composition of natural transformations. Is that on purpose?

kim-em replied:
No, it wasn't on purpose; I've rearranged now.

jcommelin commented on docs/theories/categories.md:
```diff
@@ -0,0 +1,78 @@
+### Categories
+
+We use the `⟶` (`\hom`) arrow to denote sets of morphisms, as in `X ⟶ Y`.
+This leaves the actual category implicit; it is inferred from the type of X and Y but typeclass inference.
```
s/but/by/ -- "by typeclass inference".

kim-em replied:
Fixed.

jcommelin commented on category_theory/category.lean:
```diff
@@ -0,0 +1,44 @@
+  (right_identity : ∀ {X Y : Obj} (f : Hom X Y), compose f (identity Y) = f . obviously)
+  (associativity  : ∀ {W X Y Z : Obj} (f : Hom W X) (g : Hom X Y) (h : Hom Y Z), compose (compose f g) h = compose f (compose g h) . obviously)
+
+notation `𝟙` := category.identity     -- type as \b1
```
I can imagine that we will also love to have this notation for the unit object in a monoidal category. But then... we also want good notation for identity morphisms, and I don't have a good alternative in mind.

kim-em replied:
Besides the double-struck 1, there are several other variant numeral 1s available in unicode:

"bold": 𝟏	
"sans-serif": 𝟣
"sans-serif bold": 𝟭
"monospace": 𝟷

Currently none of these have shortcuts in VSCode, but those are easy to add. My monoidal categories library hasn't fixed on any notation for monoidal units. I'm not sure that having two slightly different 1s is a good idea.

jcommelin commented on docs/theories/categories.md:
```diff
@@ -0,0 +1,78 @@
+We use `𝟙` (`\b1`) to denote identity morphisms, as in `𝟙 X`.
+
+We use `≫` (`\gg`) to denote composition of morphisms, as in `f ≫ g`, which means "`f` followed by `g`".
+This is the opposite order than the usual convention (which is lame).  
```
Mathematicians are going to love you for this [/sarcasm]. Is it possible to also have notation for "ordinary" composition? Would it even be possible to overload `\circ`? Or will madness ensue? If so, then preferably something that reminds of `\circ`.

kim-em replied:
I will investigate overloading `\circ`, or finding something nearby.

I'd been hoping that everyone would be relieved that starting to write mathematics in Lean gave them a good opportunity to abandon the no-good-terrible-horrible `\circ` notation after so many years. :-)

rwbarton replied:
Regardless of one's personal preferences, for the purpose of formalizing existing mathematics, it's greatly helpful if the Lean notation is as similar as possible to the source notation; and that will almost always require having some form of applicative-order composition notation available.

In my own project, I just write ``local notation f ` ∘ `:80 g:80 := g ≫ f`` at the top of most files, and this works pretty well for me, though I occasionally need to write `function.comp` for the usual `∘` operator. Not a very good solution for the library, though.

jcommelin commented on docs/theories/categories.md:
```diff
@@ -0,0 +1,78 @@
+
+For now we just write out `τ.components X`.
+
+For vertical and horiztonal composition of natural transformations we "cutely" use `⊟` (`\boxminus`) and `◫` (currently untypeable, but we could ask for `\boxbar`).
```
Or maybe `\boxvert`(?), which arguably has better mnemonics.

kim-em replied:
Good suggestion, I've included that.

jcommelin reviewed (commented):
Mostly trivial comments. I can't really comment on the Lean.

kim-em committed:
corrections from review

kim-em committed:
removing second ematch attribute on associativity_lemma

kim-em reviewed (commented):
Originally I thought it was inappropriate to put @[simp] on associativity, but it seems to work out okay.

Removing it here entirely changes some proofs; in some places we need to do rewrites (which can nevertheless be discovered automatically) instead of just simp,

kim-em committed:
being slightly more careful about variables

(I don't think there were any actual errors,
but I was sometimes using an argument
when there was a variable of the same
name available.)

kim-em committed:
fix(notation): transformation components

Using `@>` per @rwbarton's suggestion.

kim-em committed:
fix(notation): more conventional notation for composition

kim-em committed:
Merge branch 'master' into category_theory

johoelzl commented on category_theory/category.lean:
```diff
@@ -0,0 +1,44 @@
+import tactic.make_lemma
+import tactic.interactive
+
+namespace categories
```
what about `category_theory` instead?

johoelzl commented on category_theory/category.lean:
```diff
@@ -0,0 +1,44 @@
+-- Copyright (c) 2017 Scott Morrison. All rights reserved.
```
The header we use look like:

```lean
/-
Copyright (c) 2014 Microsoft Corporation. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Leonardo de Moura
notation, basic datatypes and type classes
-/
```

so use `/--/` instead of `--`. The difference looks minor, but might be helpful when we use scripts to analyze the header.

johoelzl commented on category_theory/functor/default.lean:
```diff
@@ -0,0 +1,75 @@
+
+open categories
+
+namespace categories.functor
```
I think everything should went into `category_theory`.

johoelzl commented on category_theory/functor/default.lean:
```diff
@@ -0,0 +1,75 @@
+structure Functor (C : Type u₁) [category.{u₁ v₁} C] (D : Type u₂) [category.{u₂ v₂} D] : Type (max u₁ v₁ u₂ v₂) :=
+  (onObjects     : C → D)
+  (onMorphisms   : Π {X Y : C}, (X ⟶ Y) → ((onObjects X) ⟶ (onObjects Y)))
+  (identities    : ∀ (X : C), onMorphisms (𝟙 X) = 𝟙 (onObjects X) . obviously)
```
Now you have `categories.functor.Functor`, I think it should be just `category_theory.functor`.

kim-em replied:
Thanks @johoelzl, I've made all the changes you suggested. I'm very happy to adjust naming to match `mathlib` conventions better, although would like to err on the side of longer names over ambiguous or not-immediately-obvious ones, where reasonable.

johoelzl reviewed (commented):
Should we really break with the naming conventions we have in Lean? I especially don't like `onObjects` and `onMorphisms`. But I would prefer if most names are a lower case with `_`.

kim-em committed:
adjusting namespaces, capitalisation, and headers

Commit:
move functor/default.lean to functor.lean

(Later PRs will add more files to the functor/ directory, but that can wait.)

Commit:
oops

Commit:
fixing indentation

Commit:
namespace for instances

Commit:
removing unnecessary `@>` notation for natural transformations

kim-em committed:
Merge commit 'a7b749f0a743ab467e1cb7d688e4a8a0f7a82874' into category_theory

Commit:
renaming, namespacing, and removing a spurious attribute

Commit:
better naming, namespaces, and coercions

Commit:
updating documentation

kim-em committed:
Merge branch 'category_theory' of github.com:semorrison/mathlib into category_theory

kim-em committed:
Merge branch 'master' into category_theory

Commit:
renaming id

Commit:
reordering definitions

Commit:
Merge branch 'category_theory' of github.com:semorrison/mathlib into category_theory

Commit:
rfl lemmas for has_one

Commit:
formatting

Commit:
formatting

Commit:
renaming: snake_case

digama0 commented on category_theory/functor.lean:
```diff
@@ -0,0 +1,87 @@
+(obj           : C → D)
+(map           : Π {X Y : C}, (X ⟶ Y) → ((obj X) ⟶ (obj Y)))
+(map_id        : ∀ (X : C), map (𝟙 X) = 𝟙 (obj X) . obviously)
+(functoriality : ∀ {X Y Z : C} (f : X ⟶ Y) (g : Y ⟶ Z), map (f ≫ g) = (map f) ≫ (map g) . obviously)
```
I think I will insist on this one. I know that literal description of axioms doesn't appeal to you, but there are just too many different ways of phrasing lemmas and it is a *huge* benefit to users that are bouncing between fields of mathematics that they may not know in detail the "traditional" way, to be able to guess names based solely on how they are written rather than what they mean. As a simple example, the symmetric version of this equality would also be called "functoriality" by a mathematician, as well as a hypothetical definition like `commutes map (≫)` which unfolds to this equality. You have `map` and `comp` being applied; so it is called `map_comp`.

digama0 commented on category_theory/functor.lean:
```diff
@@ -0,0 +1,87 @@
+{ F   := λ F, C → D,
+  coe := λ F, F.obj }
+
+@[simp] lemma unfold_obj_coercion (F : C ↝ D) (X : C) : F X = F.obj X := rfl
```
I would prefer a name like `coe_def` for this.

digama0 commented on category_theory/functor.lean:
```diff
@@ -0,0 +1,87 @@
+@[simp] lemma identity.on_objects (X : C) : (identity C) X = X := rfl
+@[simp] lemma identity.on_morphisms {X Y : C} (f : X ⟶ Y) : (identity C).map f = f := rfl
+@[simp] lemma has_one.on_objects (X : C) : (1 : C ↝ C) X = X := rfl
+@[simp] lemma has_one.on_morphisms {X Y : C} (f : X ⟶ Y) : (1 : C ↝ C).map f = f := rfl
```
Do we need both of these? I would prefer to have `identity` simplify to `1` so that we only need the second set of lemmas. Unless you are having similar problems as with simp under coe, but I don't think this should happen.

digama0 replied:
> My inclination is to actually remove entirely the `has_one` instances. I think they're more cosmetic than useful. What do you think?

What is the relation between this `identity` and the `\b1` symbol? I would like to have some kind of symbol for this which is one-like. But it doesn't hurt to have a `has_one` instance as well; just pick one or the other to be the "idiomatic" one and use it in all your statements, and have a simp lemma from any alternative ways to write it.

kim-em replied:
Done.

kim-em replied:
Ah, I missed this comment. `identity` is an "outside" identity, that is `C.identity` is the identity functor from `C` to itself. `id` (and the `\b1` symbol) are for the identities "inside" a category, e.g. `category.id X` is a morphism from the object `X` to itself.

I agree it would be nice to be clearer here. Suggestions for renamings very welcome!

digama0 replied:
How about `functor.id` (protected)? I can live with `C.id` as a notation, does that work for you?

Edit: oh wait, `C` is a category not a functor. I can still live with `functor.id C`, unless there is a good reason to further abbreviate it.

digama0 commented on category_theory/natural_transformation.lean:
```diff
@@ -0,0 +1,114 @@
+
+structure natural_transformation (F G : C ↝ D) : Type (max u₁ v₂) :=
+(components : Π X : C, (F X) ⟶ (G X))
+(naturality : ∀ {X Y : C} (f : X ⟶ Y), (F.map f) ≫ (components Y) = (components X) ≫ (G.map f) . obviously)
```
Could we perhaps abbreviate these names? `components` will appear a lot so I would suggest something with 3-5 characters, and `natural_transformation` could be abbreviated to `nat_trans` without much loss in readability. @johoelzl Thoughts? I don't even think it has to be an abbreviation of the word "components", something like `f` or `app` would work fine.

kim-em replied:
Okay, `nat_trans` coming up. I really don't understand the desire to abbreviate. Once upon a time memory was expensive, but it's not anymore, and code is so much more readable if you use unambiguous english words, particularly the explicit mathematical terms everyone is used to.

kim-em replied:
I've renamed `components` to `app`, though it makes me feel sick to obfuscate my code. :-)

digama0 replied:
When I'm staring at a 30 line goal, long names make everything harder to look at. Lean already has significant term bloat, so I try to control what I can. Symbols are especially good at making things concise and readable, but they generally aren't an option when writing proofs. The names of major concepts have to be especially short because they appear as components in theorem names, and when you use a "name the symbols" approach to theorem naming those names can get long and repetitive, so short components pay for themselves.

I notice that this definition does not have a doc comment. I would much prefer you add all explanatory information to a doc comment rather than giving everything a long name.

kim-em replied:
Okay! doc comments coming up.

digama0 commented on category_theory/natural_transformation.lean:
```diff
@@ -0,0 +1,114 @@
+variables {F G H : C ↝ D}
+
+-- We'll want to be able to prove that two natural transformations are equal if they are componentwise equal.
+@[extensionality] lemma componentwise_equal (α β : F ⟹ G) (w : ∀ X : C, α X = β X) : α = β :=
```
this theorem should be called `ext`

kim-em replied:
Done.

Commit:
renaming coe rfl lemmas

Commit:
functoriality -> map_comp

digama0 commented on tactic/make_lemma.lean:
```diff
@@ -0,0 +1,43 @@
+
+/- Make lemma takes a structure field, and makes a new, definitionally simplified copy of it, appending `_lemma` to the name.
+   The main application is to provide clean versions of structure fields that have been tagged with an auto_param. -/
+meta def make_lemma (d : declaration) (new_name : name) : tactic unit :=
```
The name of this tactic is extremely generic. How about `mk_structure_lemma` or `restate_axiom`?

kim-em replied:
Done: `restate_axiom`.

Commit:
rfl lemmas for identity C and identity F (reducing both to `1`)

digama0 reviewed (commented):
It's looking good. I think we are nearly there, just a few more nitpicks and then it will be ready.

Commit:
renaming ext lemma to `ext`

Commit:
rename `natural_transformation` to `nat_trans`

Commit:
rename `make_lemma` to `restate_axiom`

Commit:
renaming nat_trans.components to nat_trans.app

Commit:
oops, fix import

digama0 commented on category_theory/category.lean:
```diff
@@ -43,9 +43,9 @@ infixr ` ≫ `:80 := category.comp -- type as \gg
-make_lemma category.assoc
+restate_axiom category.id_comp
+restate_axiom category.comp_id
+restate_axiom category.assoc
```
missed a spot (`make_lemma`)

kim-em replied:
Fixed.

Commit:
adding doc_comments, and `protected`

Commit:
formatting

digama0 commented on category_theory/functor.lean:
```diff
@@ -49,21 +57,21 @@ namespace category
-@[simp] lemma has_one.on_objects (X : C) : (1 : C ↝ C) X = X := rfl
-@[simp] lemma has_one.on_morphisms {X Y : C} (f : X ⟶ Y) : (1 : C ↝ C).map f = f := rfl
+@[simp] protected lemma has_one.on_objects (X : C) : (1 : C ↝ C) X = X := rfl
+@[simp] protected lemma has_one.on_morphisms {X Y : C} (f : X ⟶ Y) : (1 : C ↝ C).map f = f := rfl
```
these don't all need to be protected, just `id` itself (assuming you call it `id` ;) )

kim-em replied:
So we don't have the option of calling it `id`, as there's already a structure field `category.id`, for the identity morphism on an object. This `identity` is for the identity functor on the category itself.

Commit:
removing `has_one` instances, per zulip chat, and adding a `vcomp.assoc` lemma

kim-em committed:
removing the attribute that `restate_axiom` used to add

(it was causing a problem on travis, but not locally? anyway it's useless)

digama0 commented on category_theory/functor.lean:
```diff
@@ -0,0 +1,92 @@
+variable {C}
+
+@[simp] lemma id.on_objects (X : C) : (functor.id C) X = X := rfl
+@[simp] lemma id.on_morphisms {X Y : C} (f : X ⟶ Y) : (functor.id C).map f = f := rfl
```
these should be named `id_obj` and `id_map` to match the new functor names, same with `comp` and `vcomp`, `hcomp`

kim-em replied:
Fixed, thanks!

digama0 commented on category_theory/natural_transformation.lean:
```diff
@@ -0,0 +1,117 @@
+{ app        := λ X, 𝟙 (F X),
+  naturality := begin /- `obviously'` says: -/ intros, dsimp, simp end }
+
+@[simp] lemma id.app (F : C ↝ D) (X : C) : (nat_trans.id F) X = 𝟙 (F X) := rfl
```
we usually use `_` instead of `.` for theorems referring to a definition unless it makes sense with projection notation

kim-em replied:
Fixed!

kim-em committed:
fixing names
